analyzers/src/SonarAnalyzer.CSharp/Rules/LockedFieldShouldBeReadonly.cs

```diff
@@ -0,0 +1,79 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class LockedFieldShouldBeReadonly : SonarDiagnosticAnalyzer
+{
+    private const string DiagnosticId = "S2445";
+
+    private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, "Do not lock on {0}, use a readonly field instead.");
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterNodeAction(CheckLockStatement, SyntaxKind.LockStatement);
+
+    private static void CheckLockStatement(SonarSyntaxNodeReportingContext context)
+    {
+        var expression = ((LockStatementSyntax)context.Node).Expression?.RemoveParentheses();
+        if (IsCreation(expression))
+        {
+            ReportIssue("a new instance because is a no-op");
+        }
+        else
+        {
+            var lazySymbol = new Lazy<ISymbol>(() => context.SemanticModel.GetSymbolInfo(expression).Symbol);
+            if (IsOfTypeString(expression, lazySymbol))
+            {
+                ReportIssue("strings as they can be interned");
+            }
+            else if (expression is IdentifierNameSyntax && lazySymbol.Value is ILocalSymbol lockedSymbol)
+            {
+                ReportIssue($"local variable '{lockedSymbol.Name}'");
+            }
+            else if (FieldWritable(expression, lazySymbol) is { } lockedField)
+            {
+                ReportIssue($"writable field '{lockedField.Name}'");
+            }
+        }
+
+        void ReportIssue(string message) =>
+            context.ReportIssue(Diagnostic.Create(Rule, expression.GetLocation(), message));
+    }
+
+    private static bool IsCreation(ExpressionSyntax expression) =>
+        expression.IsAnyKind(
+            SyntaxKind.ObjectCreationExpression,
+            SyntaxKind.AnonymousObjectCreationExpression,
+            SyntaxKind.ArrayCreationExpression,
+            SyntaxKind.ImplicitArrayCreationExpression,
+            SyntaxKind.QueryExpression);
+
+    private static bool IsOfTypeString(ExpressionSyntax expression, Lazy<ISymbol> lazySymbol) =>
+        expression.IsAnyKind(SyntaxKind.StringLiteralExpression, SyntaxKind.InterpolatedStringExpression)
+        || lazySymbol.Value.GetSymbolType().Is(KnownType.System_String);
+
+    private static IFieldSymbol FieldWritable(ExpressionSyntax expression, Lazy<ISymbol> lazySymbol) =>
```

Comment:
```suggestion
    private static IFieldSymbol FieldIsWritable(ExpressionSyntax expression, Lazy<ISymbol> lazySymbol) =>
```

Reply:
The name `FieldIsWritable` seems to assume that the `expression` provided as input is necessarily a field. 
In this case the method receives an `ExpressionSyntax` which may or may not be a field: for example it may be a `SimpleMemberAccessExpressionSyntax` accessing a property, for which `Expression` would be an `IPropertySymbol`.

`IsFieldWritable` would already be a better name, but imo violates the principle of least astonishment: one would expect a method starting with `Is` to return a `bool`. Instead, this method returns a `IFieldSymbol` which may or may not be null, depending on whether all the conditions of "being a member", "referring to a field" and "being a non-readonly field" are respected.

The closest semantics to this is the one of (a very elaborate) `as` operator: if you prefer, I may change the name into `AsWritableField`. `WritableField` may suggest a "cast" semantic, i.e. it may suggest throwing an exception rather than returning `null`.

However, I would avoid `IsFieldWritable` or `FieldIsWritable`... unless there is a specific code convention driving the choice. If there's not, I would assume any options of the ones above, except maybe `IsFieldWritable` due to POLA violation, should be considered valid for this PR.

Reply:
🤦  my bad - you are right.
I think I'd just change it then to `FindWritableField` or `ExtractFieldSymbol` or something a bit more descriptive if you have other ideas.
IMO just `FieldWritable` is a bit confusing.